---
title: "02-trim-primers"
output: html_document
date: "2023-04-25"
---

# Install packages and software

To get cutadapt, I downloaded the miniconda installer using curl 

install cutadapt software
```{bash}
#download miniconda from url
curl -O https://repo.anaconda.com/miniconda/Miniconda3-py38_23.3.1-0-Linux-x86_64.sh

```

and then moved to the directory it was in by typing `cd code` in the terminal. I then typed bash Miniconda3-latest-Linux-x86_64.sh in the terminal to start the setup process. Conda is located in: /home/shared/8TB_HDD_02/schulh2/miniconda3. I checked taht installation worked by typing `conda list` in the terminal. In the terminal I then typed:
`conda config --add channels bioconda
conda config --add channels conda-forge
conda create -n cutadaptenv cutadapt`

To activate the conda environment, in the terminal I typed `conda activate cutadaptenv`


install packages
```{r}
#install dada2
install.packages("BiocManager")
library(BiocManager)
BiocManager::install("dada2")
library("dada2")
```


# Make lists of forward and reverse read files

list files for forward and reverse reads
```{r}
#set file path
path <- "../data"
#make matched list of forward and reverse reads
fnFs <- sort(list.files(path, pattern = "_R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2_001.fastq.gz", full.names = TRUE))

#specify primer sequences
FWD <- "GGWACWGGWTGAACWGTWTAYCCYCC"  
REV <- "TAIACYTCIGGRTGICCRAARAAYCA"  
```


# Check if primers are actually in data

get orientations of primers
```{r}
allOrients <- function(primer) {
    # Create all orientations of the input sequence
    require(Biostrings)
    dna <- DNAString(primer)  # The Biostrings works w/ DNAString objects rather than character vectors
    orients <- c(Forward = dna, Complement = complement(dna), Reverse = reverse(dna), 
        RevComp = reverseComplement(dna))
    return(sapply(orients, toString))  # Convert back to character vector
}
FWD.orients <- allOrients(FWD)
REV.orients <- allOrients(REV)
FWD.orients
```

pre-filter sequences to remove those with N
```{r}
fnFs.filtN <- file.path(path, "filtN", basename(fnFs)) # Put N-filterd files in filtN/ subdirectory
fnRs.filtN <- file.path(path, "filtN", basename(fnRs))


filterAndTrim(fnFs, fnFs.filtN, fnRs, fnRs.filtN, maxN = 0, multithread = TRUE)

```

count number of times primers occur
```{r}
primerHits <- function(primer, fn) {
    # Counts number of reads in which the primer is found
    nhits <- vcountPattern(primer, sread(readFastq(fn)), fixed = FALSE)
    return(sum(nhits > 0))
}
rbind(FWD.ForwardReads = sapply(FWD.orients, primerHits, fn = fnFs.filtN[[1]]), 
    FWD.ReverseReads = sapply(FWD.orients, primerHits, fn = fnRs.filtN[[1]]), 
    REV.ForwardReads = sapply(REV.orients, primerHits, fn = fnFs.filtN[[1]]), 
    REV.ReverseReads = sapply(REV.orients, primerHits, fn = fnRs.filtN[[1]]))
```


check if files have forward primer sequence
```{bash}
cd ../data
grep -r GGAACTGGTTGAACTGTATATCCTCC .
```

take a look at data

```{bash}
cd ../data
zcat 2018APRP402ObB-COI-59samples_S191_L001_R1_001.fastq.gz | head -n 2
```

# Remove Primers

Tell R the path to the cutadapt command

```{r}
cutadapt <- "/home/shared/8TB_HDD_02/schulh2/miniconda3/envs/cutadaptenv/bin/cutadapt" 
system2(cutadapt, args = "--version") 
```


```{r}
path.cut <- file.path(path, "cutadapt")
if(!dir.exists(path.cut)) dir.create(path.cut)
fnFs.cut <- file.path(path.cut, basename(fnFs))
fnRs.cut <- file.path(path.cut, basename(fnRs))

FWD.RC <- dada2:::rc(FWD)
REV.RC <- dada2:::rc(REV)
# Trim FWD and the reverse-complement of REV off of R1 (forward reads)
R1.flags <- paste("-g", FWD, "-a", REV.RC) 
# Trim REV and the reverse-complement of FWD off of R2 (reverse reads)
R2.flags <- paste("-G", REV, "-A", FWD.RC) 
# Run Cutadapt
for(i in seq_along(fnFs)) {
  system2(cutadapt, args = c(R1.flags, R2.flags, "-n", 2, # -n 2 required to remove FWD and REV from reads
                             "-o", fnFs.cut[i], "-p", fnRs.cut[i], # output files
                             fnFs.filtN[i], fnRs.filtN[i])) # input files
}

```

