---
title: "03-filter-reads"
output: html_document
date: "2023-05-04"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,         # Display code chunks
  eval = FALSE,        # Evaluate code chunks
  warning = FALSE,     # Hide warnings
  message = FALSE,     # Hide messages
  fig.width = 6,       # Set plot width in inches
  fig.height = 4,      # Set plot height in inches
  fig.align = "center" # Align plots to the center
)
```
load packages
```{r, eval=TRUE}
library(BiocManager)
library(ShortRead)
library("dada2")
```

# Inspect read quality

Get sample names
```{r, eval=TRUE}
# Forward and reverse fastq filenames have the format:
cutFs <- sort(list.files(path.cut, pattern = "_R1_001.fastq.gz", full.names = TRUE))
cutRs <- sort(list.files(path.cut, pattern = "_R2_001.fastq.gz", full.names = TRUE))

# Extract sample names, assuming filenames have format:
get.sample.name <- function(fname) strsplit(basename(fname), "_")[[1]][1]
sample.names <- unname(sapply(cutFs, get.sample.name))
head(sample.names)
```

plot forward read quality for first two files
```{r, eval=TRUE}
plotQualityProfile(cutFs[1:2])
```

plot reverse read quality for first two files
```{r, eval=TRUE}
plotQualityProfile(cutRs[1:2])
```
The forward reads are better quality, as we might expect.


```{r, eval=TRUE}
filtFs <- file.path(path.cut, "filtered", basename(cutFs))
filtRs <- file.path(path.cut, "filtered", basename(cutRs))
```

```{r, eval=TRUE}
# get sample names to see which files match
get.sample.name <- function(fname) strsplit(basename(fname), "_")[[1]][1]
sample.namesF <- unname(sapply(cutFs, get.sample.name))
sample.namesR <- unname(sapply(cutRs, get.sample.name))
```

Check for differences between lists and remove files that don't have a matching pair
```{r, eval=TRUE}
difsR <- setdiff(sample.namesR,sample.namesF)
difsR
difsF <- setdiff(sample.namesF,sample.namesR)
difsF

```
It looks like there are no differences, so we can proceed


## filter reads using quality thresholds
maxN: allows zero Ns
max EE: set to 5 for forward and reverse
truncQ: set to 2
minlen: minimum length is 100 reads
rm.phix:remove phix
this step takes a long time
```{r}
out <- filterAndTrim(cutFs, filtFs, cutRs, filtRs, maxN = 0, maxEE = c(5, 5), 
    truncQ = 2, minLen = 100, rm.phix = TRUE, compress = TRUE, multithread = TRUE) 
```

Check how many samples went through filtering
```{bash, eval=TRUE}
cd /Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/data/fastq-files/cutadapt

ls | wc -l

cd /Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/data/fastq-files/cutadapt/filtered

ls | wc -l
```

learn the error rates

forward reads
```{r, eval=TRUE, cache=TRUE}
errF <- learnErrors(filtFs, multithread=TRUE)
```
reverse reads
```{r, eval=TRUE, cache=TRUE}
errR <- learnErrors(filtRs, multithread=TRUE)
```

plot the errors
```{r, eval=TRUE}
plotErrors(errF, nominalQ=TRUE)
#looks okay! good to proceed
```
```{r, eval=TRUE}
plotErrors(errR, nominalQ=TRUE)
#looks okay! good to proceed
```

QC check on files after filtering were trimmed

use fastqc 
```{r, engine="bash"}
cd /Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/data/fastq-files/cutadapt/filtered
/Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/software/FastQC/fastqc *fastq.gz -o /Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/output/after-filtering-qc
```

make multiqc report to check files after trimming primers
```{r, engine="bash"}
cd /Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/output/after-filtering-qc

multiqc .
```

![screenshot of multiqc mean quality scores after filtering](/Users/hailaschultz/GitHub/haila-ZoopMetabarcoding/screenshots/after-filtering_multiqc.png)