---
title: "Research Project Presentation"
author: Haila Schultz
format: 
  revealjs:
    theme: simple
    scrollable: true
editor: visual
---

## Dataset

::: incremental
-   Washington Ocean Acidification Center Cruises
-   Paired zooplankton samples from net tows: morphology and metabarcoding
-   metabarcoding marker: mitochondrial cytochrome C oxidase subunit I (COI)
:::

## Number of Samples

| Month | 2019 | 2020 | 2020 |
|-------|------|------|------|
| April | 7    | 7    | 7    |
| July  | 7    | 7    | 7    |
| Sept  | 7    | 7    | 7    |

## Stations

![](WOAC-station.png)

## Project Goals

::: incremental
1.  Taxonomically ID zooplankton communities using sequence reads

2.  Compare zooplankton communities among

-   years
-   seasons
-   locations
:::

## Methods to-date

1.  Download data
2.  Trim primers using cutadapt (in progress)

## Trimming primers

```{r, eval=FALSE}
#| echo: true
#| code-line-numbers: "7-9"
#set file path
path <- "../data"
#make matched list of forward and reverse reads
fnFs <- sort(list.files(path, pattern = "_R1_001.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern = "_R2_001.fastq.gz", full.names = TRUE))

#specify primer sequences
FWD <- "GGWACWGGWTGAACWGTWTAYCCYCC"  
REV <- "TAIACYTCIGGRTGICCRAARAAYCA"  
```

## Take a look at one file

```{r, engine='bash', echo=TRUE, message=FALSE}
cd ../data
zcat 2018APRP402ObB-COI-59samples_S191_L001_R1_001.fastq.gz | head -n 2
```

## Methods for the next four weeks

::: incremental
1.  Filter reads based on quality using DADA2
2.  Merge forward and reverse reads
3.  Run BLASTn against the NCBI nt database as a reference
4.  Calculate diversity index for each sample
5.  Make NMDS plots colored by year, month, and location
6.  Make barplots to compare relative abundances of taxa among samples
:::
